python/dgl/kernel.py

```diff
@@ -0,0 +1,260 @@
+"""Module for dgl kernels for graph computation."""
+from __future__ import absolute_import
+
+from ._ffi.function import _init_api
+from .ndarray import empty
+
+def infer_binary_feature_shape(lhs, rhs):
+    """Infer the output feature shape after a binary operation between lhs and rhs.
+
+    Parameter
+    ---------
+    lhs : dgl.ndarray.NDArray
+        The lhs tensor.
+    rhs : dgl.ndarray.NDArray
+        The rhs tensor.
+
+    Returns
+    -------
+    tuple of int
+        The output feature shape.
+    """
+    ret = _CAPI_DGLKernelInferBinaryFeatureShape(lhs, rhs)
+    return tuple(ret.asnumpy())
+
+def binary_op_reduce(reducer, binary_op, graph, lhs, rhs, lhs_data, rhs_data,
+                     out_data, lhs_mapping=None, rhs_mapping=None,
+                     out_mapping=None):
+    """Perform binary operation between the given data and reduce by the graph.
+
+    Broadcasting is supported for feature dimensions.
+
+    Optional id mapping arrays could be provided to read/write from/to locations
+    other than node/edge ids.
+
+    Parameter
+    ---------
+    reducer : str
+        The type of the reducer ("sum", "max", "min", "mean", "prod", "none").
+        If the reducer is "none", the output is an edge feature tensor.
+        Otherwise, a node feature tensor is returned.
+    binary_op : str
+        The type of the binary functor ("add", "mul", "sub", "div").
+    graph : GraphIndex
+        The graph
+    lhs : int
+        The lhs target (src, dst, edge)
+    rhs : int
+        The rhs target (src, dst, edge)
+    lhs_data : NDArray
+        The lhs data.
+    rhs_data : NDArray
+        The rhs data.
+    out_data : NDArray
+        The out data.
+    lhs_mapping : NDArray
+        The lhs id mapping array.
+    rhs_mapping : NDArray
+        The rhs id mapping array.
+    out_mapping : NDArray
+        The out id mapping array.
```

Comment:
If `xxx_mapping` are given, I assume that `lhs`, `rhs` and `out` must stay consistent?  Like
* If `reducer, lhs, rhs = none, src, edge`, then `rhs_mapping == out_mapping && lhs_mapping == src_nodes_of(rhs_mapping)` must hold.
* If `reducer, lhs, rhs = sum, dst, edge`, then `out_mapping == dst_nodes_of(rhs_mapping) == lhs_mapping` must hold.

Reply:
I don't quite get this one...

lhs_mapping is for lhs, rhs_mapping is for rhs, out_mapping is for out. Whenever we want to access XXX, we need to go through XXX_mapping first. It's just one layer of indirection.

Reply:
Added some description here. @BarclayII see if looks better for you.

Reply:
I see.  So essentially `xxx_mapping` are actually slicing the lhs/rhs/output data to match the nodes/edges of `graph`.  And `graph` could be either the entire original graph or a part of it as in `send_and_recv`.